File: src/formspec/fel/environment.py
"""FEL evaluation environment -- the data layer beneath the evaluator.

Resolves ``$field`` paths against primary instance data, manages a push/pop
scope stack for ``let``-bindings, provides ``RepeatContext`` for
``@current``/``@index``/``@count``, serves ``MipState`` lookups for
``valid()``/``relevant()``/``readonly()``/``required()``, and exposes named
secondary instances via ``@instance('name')``.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .types import FelNull, FelValue, FelObject, from_python, is_null


@dataclass
class RepeatContext:
    """Per-iteration state for repeat groups: backs @current, @index (1-based), @count, parent(), and prev()/next() navigation."""
    current: FelValue  # The current row object
    index: int  # 1-based
    count: int
    parent: FelValue  # Parent context
    collection: list  # All rows


@dataclass
class MipState:
    """XForms-style Model Item Properties for a single field, queried by valid()/relevant()/etc. special forms."""
    valid: bool = True
    relevant: bool = True
    readonly: bool = False
    required: bool = False


class Environment:
    """Evaluation context providing field resolution, lexical scoping, repeat iteration, and MIP state.

    Primary data dict backs ``$field`` refs; scope stack backs ``let``-bindings;
    ``RepeatContext`` backs ``@current``/``@index``/``@count``; named instances
    back ``@instance('name')``.
    """

    def __init__(
        self,
        data: dict | None = None,
        instances: dict[str, dict] | None = None,
        mip_states: dict[str, MipState] | None = None,
        variables: dict[str, 'FelValue'] | None = None,
    ):
        """Set up with primary instance data, optional named secondary instances, optional per-field MIP states, and optional pre-computed named variables."""
        self.data = data or {}
        self.instances = instances or {}
        self.mip_states = mip_states or {}
        self.variables = variables or {}
        self._scope_stack: list[dict[str, FelValue]] = []
        self.repeat_context: RepeatContext | None = None

    def push_scope(self, bindings: dict[str, FelValue]) -> None:
        """Push a lexical scope frame. Key '' rebinds bare ``$`` (used by countWhere predicates)."""
        self._scope_stack.append(bindings)

    def pop_scope(self) -> None:
        """Pop the most recent lexical scope."""
        self._scope_stack.pop()

    def lookup_let_binding(self, name: str) -> FelValue | None:
        """Search scope stack innermost-first for a let-bound variable. Returns None if unbound."""
        for scope in reversed(self._scope_stack):
            if name in scope:
                return scope[name]
        return None

    def resolve_field(self, path: list[str]) -> FelValue:
        """Resolve a ``$field`` path through the resolution chain.

        Empty path (bare ``$``): scope stack -> repeat current -> full data dict.
        Non-empty: first segment checked against let-bindings, then walks primary
        instance data. Returns FelNull for missing paths.
        """
        if not path:
            # Bare $ -- check scope stack first (countWhere binds $ per-element)
            dollar_val = self.lookup_let_binding('')
            if dollar_val is not None:
                return dollar_val
            # Then repeat context
            if self.repeat_context:
                return self.repeat_context.current
            return from_python(self.data)

        # Check let-bindings first (for first segment only)
        let_val = self.lookup_let_binding(path[0])
        if let_val is not None:
            return self._resolve_tail(let_val, path[1:])

        # Resolve against instance data
        return self._resolve_path(self.data, path)

    def resolve_context(self, name: str, arg: str | None, tail: list[str]) -> FelValue:
        """Resolve ``@name(arg).tail`` context references.

        Supported: @current/.tail (repeat row), @index/@count (repeat metadata),
        @instance('name')/.tail (secondary data), @source/@target (mapping DSL),
        @name (definition variable lookup via the ``variables`` dict).
        Built-in names (current, index, count, instance, source, target) always take
        precedence — do not use them as variable keys.
        Unknown names with no matching variable -> FelNull.
        """
        if name == 'current':
            if self.repeat_context is None:
                return FelNull
            val = self.repeat_context.current
            return self._resolve_tail(val, tail)
        if name == 'index':
            if self.repeat_context is None:
                return FelNull
            from .types import FelNumber, fel_decimal
            return FelNumber(fel_decimal(self.repeat_context.index))
        if name == 'count':
            if self.repeat_context is None:
                return FelNull
            from .types import FelNumber, fel_decimal
            return FelNumber(fel_decimal(self.repeat_context.count))
        if name == 'instance':
            if arg is None or arg not in self.instances:
                return FelNull
            return self._resolve_path(self.instances[arg], tail)
        if name in ('source', 'target'):
            # Mapping DSL context references
            return self._resolve_path(self.data, [name] + tail)
        if name in self.variables:
            return self._resolve_tail(self.variables[name], tail)
        return FelNull

    def _resolve_path(self, obj: Any, path: list[str]) -> FelValue:
        """Walk nested dicts by string keys, converting the leaf to FelValue via from_python(). Missing key at any level -> FelNull."""
        current = obj
        for segment in path:
            if isinstance(current, dict):
                current = current.get(segment)
                if current is None:
                    return FelNull
            else:
                return FelNull
        return from_python(current)

    def _resolve_tail(self, val: FelValue, tail: list[str]) -> FelValue:
        """Walk remaining dot segments on a FelObject chain. Non-object or missing key -> FelNull."""
        if not tail:
            return val
        if isinstance(val, FelObject):
            current = val
            for seg in tail:
                if isinstance(current, FelObject) and seg in current.fields:
                    current = current.fields[seg]
                else:
                    return FelNull
            return current
        return FelNull

File: crates/fel-core/src/environment.rs
//! FEL evaluation environment with field resolution, repeats, MIP state, and instances.
//!
//! Provides `FormspecEnvironment`, a concrete `Environment` impl backed by
//! nested data dicts, repeat context, MIP states, named instances, and variables.
//!
//! Helpers such as `project_repeat_field` resolve repeat-group keys into projected field values.
#![allow(clippy::missing_docs_in_private_items)]
use rust_decimal::Decimal;
use std::collections::HashMap;

use crate::evaluator::Environment;
use crate::types::{FelDate, FelValue, parse_datetime_literal};

// ── Data structures ─────────────────────────────────────────────

/// Repeat-group iteration context (§4.3).
#[derive(Debug, Clone)]
pub struct RepeatContext {
    /// The current row value.
    pub current: FelValue,
    /// 1-based index within the repeat group.
    pub index: usize,
    /// Total instance count.
    pub count: usize,
    /// Parent repeat context (for nested repeats).
    pub parent: Option<Box<RepeatContext>>,
    /// All rows in the collection (for prev/next navigation).
    pub collection: Vec<FelValue>,
}

/// XForms Model Item Properties for a single field path.
#[derive(Debug, Clone)]
pub struct MipState {
    /// `valid($path)` result when set for this path.
    pub valid: bool,
    /// `relevant($path)`.
    pub relevant: bool,
    /// `readonly($path)`.
    pub readonly: bool,
    /// `required($path)`.
    pub required: bool,
}

#[allow(missing_docs)]
impl Default for MipState {
    fn default() -> Self {
        Self {
            valid: true,
            relevant: true,
            readonly: false,
            required: false,
        }
    }
}

// ── FormspecEnvironment ─────────────────────────────────────────

/// A full-featured environment for FEL evaluation within a Formspec engine.
///
/// Supports:
/// - Field resolution via `$field.path` (walks nested data dict)
/// - Named instances via `@instance('name')`
/// - Repeat context via `@current`, `@index`, `@count`
/// - MIP state queries via `valid()`, `relevant()`, etc.
/// - Definition variables via `@variableName`
/// - Mapping context via `@source`, `@target`
pub struct FormspecEnvironment {
    /// Primary data dict — backs `$field` references.
    pub data: HashMap<String, FelValue>,
    /// Named secondary instances — backs `@instance('name')`.
    pub instances: HashMap<String, FelValue>,
    /// MIP states per dotted field path.
    pub mip_states: HashMap<String, MipState>,
    /// Definition variables — backs `@variableName`.
    pub variables: HashMap<String, FelValue>,
    /// Current repeat context (if inside a repeat iteration).
    pub repeat_context: Option<RepeatContext>,
    /// Current runtime date for today()/now().
    pub current_datetime: Option<FelDate>,
}

impl FormspecEnvironment {
    /// Empty environment (no data, instances, or repeat context).
    pub fn new() -> Self {
        Self {
            data: HashMap::new(),
            instances: HashMap::new(),
            mip_states: HashMap::new(),
            variables: HashMap::new(),
            repeat_context: None,
            current_datetime: None,
        }
    }

    /// Set a field value by dotted path (e.g., "address.city").
    pub fn set_field(&mut self, path: &str, value: FelValue) {
        self.data.insert(path.to_string(), value);
    }

    /// Set a named instance.
    pub fn set_instance(&mut self, name: &str, value: FelValue) {
        self.instances.insert(name.to_string(), value);
    }

    /// Set MIP state for a field path.
    pub fn set_mip(&mut self, path: &str, state: MipState) {
        self.mip_states.insert(path.to_string(), state);
    }

    /// Set a variable value.
    pub fn set_variable(&mut self, name: &str, value: FelValue) {
        self.variables.insert(name.to_string(), value);
    }

    /// Set the current runtime datetime from an ISO string.
    pub fn set_now_from_iso(&mut self, iso: &str) {
        let candidate = if iso.starts_with('@') {
            iso.to_string()
        } else {
            format!("@{iso}")
        };
        self.current_datetime = parse_datetime_literal(&candidate);
    }

    /// Enter a repeat context.
    pub fn push_repeat(
        &mut self,
        current: FelValue,
        index: usize,
        count: usize,
        collection: Vec<FelValue>,
    ) {
        let parent = self.repeat_context.take().map(Box::new);
        self.repeat_context = Some(RepeatContext {
            current,
            index,
            count,
            parent,
            collection,
        });
    }

    /// Leave the current repeat context, restoring the parent.
    pub fn pop_repeat(&mut self) {
        if let Some(ctx) = self.repeat_context.take() {
            self.repeat_context = ctx.parent.map(|b| *b);
        }
    }
}

#[allow(missing_docs)]
impl Default for FormspecEnvironment {
    fn default() -> Self {
        Self::new()
    }
}

/// Walk a nested value by string path segments.
fn resolve_path(val: &FelValue, segments: &[String]) -> FelValue {
    let mut current = val.clone();
    for seg in segments {
        match &current {
            FelValue::Object(entries) => match entries.iter().find(|(k, _)| k == seg) {
                Some((_, v)) => current = v.clone(),
                None => return FelValue::Null,
            },
            FelValue::Array(entries) => {
                current = FelValue::Array(
                    entries
                        .iter()
                        .map(|entry| match entry {
                            FelValue::Object(fields) => fields
                                .iter()
                                .find(|(k, _)| k == seg)
                                .map(|(_, v)| v.clone())
                                .unwrap_or(FelValue::Null),
                            _ => FelValue::Null,
                        })
                        .collect(),
                );
            }
            _ => return FelValue::Null,
        }
    }
    current
}

fn project_repeat_field(data: &HashMap<String, FelValue>, segments: &[String]) -> Option<FelValue> {
    if segments.len() < 2 {
        return None;
    }

    let prefix = format!("{}[", segments[0]);
    let suffix = format!(".{}", segments[1..].join("."));
    let mut projected = Vec::new();

    for (key, value) in data {
        let Some(rest) = key.strip_prefix(&prefix) else {
            continue;
        };
        let Some((idx, tail)) = rest.split_once(']') else {
            continue;
        };
        if tail != suffix {
            continue;
        }
        let Ok(index) = idx.parse::<usize>() else {
            continue;
        };
        projected.push((index, value.clone()));
    }

    if projected.is_empty() {
        return None;
    }

    projected.sort_by_key(|(index, _)| *index);
    Some(FelValue::Array(
        projected.into_iter().map(|(_, value)| value).collect(),
    ))
}

#[allow(missing_docs)]
impl Environment for FormspecEnvironment {
    fn resolve_field(&self, segments: &[String]) -> FelValue {
        if segments.is_empty() {
            // Bare $ — return repeat context current, data[""], or null
            if let Some(ctx) = &self.repeat_context {
                return ctx.current.clone();
            }
            if let Some(val) = self.data.get("") {
                return val.clone();
            }
            return FelValue::Null;
        }

        let key = segments.join(".");
        // Check flat lookup first (handles dotted keys like "address.city")
        if let Some(val) = self.data.get(&key) {
            return val.clone();
        }
        // Check first segment then walk
        if let Some(val) = self.data.get(&segments[0]) {
            if segments.len() == 1 {
                return val.clone();
            }
            return resolve_path(val, &segments[1..]);
        }
        if let Some(projected) = project_repeat_field(&self.data, segments) {
            return projected;
        }
        FelValue::Null
    }

    fn resolve_context(&self, name: &str, arg: Option<&str>, tail: &[String]) -> FelValue {
        match name {
            "current" => {
                if let Some(ctx) = &self.repeat_context {
                    let base = ctx.current.clone();
                    if tail.is_empty() {
                        base
                    } else {
                        resolve_path(&base, tail)
                    }
                } else {
                    FelValue::Null
                }
            }
            "index" => {
                if let Some(ctx) = &self.repeat_context {
                    FelValue::Number(Decimal::from(ctx.index as i64))
                } else {
                    FelValue::Null
                }
            }
            "count" => {
                if let Some(ctx) = &self.repeat_context {
                    FelValue::Number(Decimal::from(ctx.count as i64))
                } else {
                    FelValue::Null
                }
            }
            "instance" => {
                if let Some(inst_name) = arg {
                    if let Some(val) = self.instances.get(inst_name) {
                        if tail.is_empty() {
                            val.clone()
                        } else {
                            resolve_path(val, tail)
                        }
                    } else {
                        FelValue::Null
                    }
                } else {
                    FelValue::Null
                }
            }
            // Definition variables: @variableName
            _ => {
                if let Some(val) = self.variables.get(name) {
                    if tail.is_empty() {
                        val.clone()
                    } else {
                        resolve_path(val, tail)
                    }
                } else {
                    FelValue::Null
                }
            }
        }
    }

    fn mip_valid(&self, path: &[String]) -> FelValue {
        let key = path.join(".");
        FelValue::Boolean(self.mip_states.get(&key).is_none_or(|s| s.valid))
    }

    fn mip_relevant(&self, path: &[String]) -> FelValue {
        let key = path.join(".");
        FelValue::Boolean(self.mip_states.get(&key).is_none_or(|s| s.relevant))
    }

    fn mip_readonly(&self, path: &[String]) -> FelValue {
        let key = path.join(".");
        FelValue::Boolean(self.mip_states.get(&key).is_some_and(|s| s.readonly))
    }

    fn mip_required(&self, path: &[String]) -> FelValue {
        let key = path.join(".");
        FelValue::Boolean(self.mip_states.get(&key).is_some_and(|s| s.required))
    }

    fn repeat_prev(&self) -> FelValue {
        if let Some(ctx) = &self.repeat_context
            && ctx.index > 1
        {
            return ctx
                .collection
                .get(ctx.index - 2)
                .cloned()
                .unwrap_or(FelValue::Null);
        }
        FelValue::Null
    }

    fn repeat_next(&self) -> FelValue {
        if let Some(ctx) = &self.repeat_context
            && ctx.index < ctx.count
        {
            return ctx
                .collection
                .get(ctx.index)
                .cloned()
                .unwrap_or(FelValue::Null);
        }
        FelValue::Null
    }

    fn repeat_parent(&self) -> FelValue {
        if let Some(ctx) = &self.repeat_context
            && let Some(parent) = &ctx.parent
        {
            return parent.current.clone();
        }
        FelValue::Null
    }

    fn current_date(&self) -> Option<FelDate> {
        self.current_datetime.as_ref().map(|dt| FelDate::Date {
            year: dt.year(),
            month: dt.month(),
            day: dt.day(),
        })
    }

    fn current_datetime(&self) -> Option<FelDate> {
        self.current_datetime.clone()
    }
}

#[cfg(test)]
mod tests {
    #![allow(clippy::missing_docs_in_private_items)]
    use super::*;

    fn s(v: &str) -> FelValue {
        FelValue::String(v.to_string())
    }

    fn num(n: i64) -> FelValue {
        FelValue::Number(Decimal::from(n))
    }

    #[test]
    fn test_basic_field_resolution() {
        let mut env = FormspecEnvironment::new();
        env.set_field("name", s("Alice"));
        env.set_field("age", num(30));

        assert_eq!(env.resolve_field(&["name".into()]), s("Alice"));
        assert_eq!(env.resolve_field(&["age".into()]), num(30));
        assert_eq!(env.resolve_field(&["missing".into()]), FelValue::Null);
    }

    #[test]
    fn test_nested_field_resolution() {
        let mut env = FormspecEnvironment::new();
        let addr = FelValue::Object(vec![
            ("city".to_string(), s("NYC")),
            ("zip".to_string(), s("10001")),
        ]);
        env.set_field("address", addr);

        assert_eq!(
            env.resolve_field(&["address".into(), "city".into()]),
            s("NYC")
        );
        assert_eq!(
            env.resolve_field(&["address".into(), "missing".into()]),
            FelValue::Null
        );
    }

    #[test]
    fn test_repeat_field_projection_from_flat_rows() {
        let mut env = FormspecEnvironment::new();
        env.set_field("rows[0].score", num(80));
        env.set_field("rows[1].score", num(30));
        env.set_field("rows[2].score", num(50));

        assert_eq!(
            env.resolve_field(&["rows".into(), "score".into()]),
            FelValue::Array(vec![num(80), num(30), num(50)]),
        );
    }

    #[test]
    fn test_repeat_context() {
        let mut env = FormspecEnvironment::new();
        let items = vec![num(10), num(20), num(30)];
        env.push_repeat(num(20), 2, 3, items);

        // @current
        assert_eq!(env.resolve_context("current", None, &[]), num(20));
        // @index (1-based)
        assert_eq!(env.resolve_context("index", None, &[]), num(2));
        // @count
        assert_eq!(env.resolve_context("count", None, &[]), num(3));
        // prev() — item at index 1 (0-indexed: 0)
        assert_eq!(env.repeat_prev(), num(10));
        // next() — item at index 3 (0-indexed: 2)
        assert_eq!(env.repeat_next(), num(30));
    }

    #[test]
    fn test_nested_repeat_context() {
        let mut env = FormspecEnvironment::new();
        let outer = vec![s("A"), s("B")];
        env.push_repeat(s("A"), 1, 2, outer);

        let inner = vec![num(1), num(2), num(3)];
        env.push_repeat(num(2), 2, 3, inner);

        // Inner context active
        assert_eq!(env.resolve_context("current", None, &[]), num(2));
        // parent() returns outer current
        assert_eq!(env.repeat_parent(), s("A"));

        env.pop_repeat();
        // Outer context restored
        assert_eq!(env.resolve_context("current", None, &[]), s("A"));
    }

    #[test]
    fn test_named_instances() {
        let mut env = FormspecEnvironment::new();
        let config = FelValue::Object(vec![("maxRetries".to_string(), num(3))]);
        env.set_instance("config", config);

        assert_eq!(
            env.resolve_context("instance", Some("config"), &["maxRetries".into()]),
            num(3)
        );
        assert_eq!(
            env.resolve_context("instance", Some("missing"), &[]),
            FelValue::Null
        );
    }

    #[test]
    fn test_variables() {
        let mut env = FormspecEnvironment::new();
        env.set_variable("total", num(100));

        assert_eq!(env.resolve_context("total", None, &[]), num(100));
    }

    #[test]
    fn test_mip_states() {
        let mut env = FormspecEnvironment::new();
        env.set_mip(
            "email",
            MipState {
                valid: false,
                relevant: true,
                readonly: false,
                required: true,
            },
        );

        assert_eq!(env.mip_valid(&["email".into()]), FelValue::Boolean(false));
        assert_eq!(env.mip_relevant(&["email".into()]), FelValue::Boolean(true));
        assert_eq!(
            env.mip_readonly(&["email".into()]),
            FelValue::Boolean(false)
        );
        assert_eq!(env.mip_required(&["email".into()]), FelValue::Boolean(true));

        // Default MIP for unknown fields
        assert_eq!(env.mip_valid(&["unknown".into()]), FelValue::Boolean(true));
        assert_eq!(
            env.mip_required(&["unknown".into()]),
            FelValue::Boolean(false)
        );
    }

    #[test]
    fn test_bare_dollar_in_repeat() {
        let mut env = FormspecEnvironment::new();
        let items = vec![num(10), num(20)];
        env.push_repeat(num(10), 1, 2, items);

        // Bare $ resolves to repeat context current
        assert_eq!(env.resolve_field(&[]), num(10));
    }
}
